benchmark/Cpp/Savina/src/micro/ThreadRing.lf

```diff
@@ -0,0 +1,150 @@
+    
+    reaction(inPrevReactor) -> outNextReactor, finished {=
+        if (inPrevReactor.get()->hasNext()) {
+            outNextReactor.set(inPrevReactor.get()->next());
+        } else {
+            finished.set();
+            reactor::log::Debug() << "Finished with count " << inPrevReactor.get()->getPingsLeft();
+        }
+    =}
+    
+    reaction(start) -> outNextReactor, finished {=
+        if (start.get()->hasNext()) {
+            outNextReactor.set(start.get()->next());
+            reactor::log::Debug() << "Starting with count " << start.get()->getPingsLeft();
+        } else {
+            finished.set();
+        }
+    =}
+}
+
+reactor ThreadRingReactorLoopOpener {
+    
+    public preamble {=
+        #include "reactor-cpp/logging.hh"
+    =}
+    
+    output outNextReactor:{=PingMessage=};
+    input inPrevReactor:{=PingMessage=};
+    input start:{=PingMessage=};
+    output finished:void;
+    
+    logical action sendToNextReactor:{=PingMessage=};
+    
+    reaction(sendToNextReactor) -> outNextReactor {=
+        outNextReactor.set(sendToNextReactor.get());
+    =}
+    
+    reaction(inPrevReactor) -> sendToNextReactor, finished {=
+        if (inPrevReactor.get()->hasNext()) {
+            sendToNextReactor.schedule(inPrevReactor.get()->next());
+        } else {
+            finished.set();
+            reactor::log::Debug() << "Finished with count " << inPrevReactor.get()->getPingsLeft();
+        }
+    =}
+    
+    reaction(start) -> sendToNextReactor, finished {=
+        reactor::log::Debug() << "Starting with count " << start.get()->getPingsLeft();
+        if (start.get()->hasNext()) {
+            sendToNextReactor.schedule(start.get()->next());
+        } else {
+            finished.set();
+        }
+    =}
+}
+
+reactor Initializer(numReactors:size_t(100), numPings:int(100000)) {
+    
+    input inStart:void;
+    output outFinished:void;
+    
+    // benchmark internal start and finish
+    input[numReactors] inFinished:void;
+    output outStart: {=PingMessage=};
+    
+    reaction(inStart) -> outStart {=
+        outStart.set(numPings);
+    =}
+    
+    reaction(inFinished) -> outFinished {=
+        outFinished.set();
+    =}
+}
+
+main reactor (numIterations:int(12), numPings:int(100000), numReactors:size_t(100)) {
+    
+    init = new Initializer(numReactors=numReactors, numPings=numPings);
+    runner = new BenchmarkRunner(numIterations=numIterations);
+    
+    runner.start -> init.inStart;
+    init.outFinished -> runner.finished;
+    
+    reaction(startup) {=
+        printBenchmarkInfo("ThreadRingBenchmark");
+        printArgs("numIterations", numIterations, "numPings", numPings, "numReactors", numReactors);
+        printSystemInfo();
+    =}
+    
+    loopOpener = new ThreadRingReactorLoopOpener();
+
+    workers = new[{=numReactors-1=}] ThreadRingReactor();
+
+    loopOpener.outNextReactor, workers.outNextReactor -> workers.inPrevReactor, loopOpener.inPrevReactor;
```

Comment:
Wouldn't it better to first create a zero-delay chain of workers and then, in a separate statement, connect the last worker to the first using an `after 0` clause? That would obviate the need for the `ThreadRingReactorLoopOpener` altogether.

Reply:
Yes, I think this is a relict of the times before we had that connection syntax. I can change it.